File: scripts/solar/solar-common/controller/AutoMissileAttackController.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__ArraySort = ____lualib.__TS__ArraySort
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 1,["9"] = 1,["10"] = 2,["11"] = 2,["12"] = 3,["13"] = 3,["14"] = 4,["15"] = 4,["16"] = 5,["17"] = 5,["18"] = 6,["19"] = 6,["20"] = 7,["21"] = 7,["22"] = 8,["23"] = 8,["24"] = 25,["25"] = 25,["26"] = 25,["27"] = 67,["28"] = 67,["29"] = 67,["31"] = 29,["32"] = 31,["33"] = 32,["34"] = 34,["35"] = 35,["36"] = 41,["37"] = 43,["38"] = 44,["39"] = 45,["40"] = 47,["41"] = 48,["42"] = 49,["43"] = 50,["44"] = 59,["45"] = 60,["46"] = 62,["47"] = 68,["48"] = 69,["49"] = 70,["50"] = 71,["51"] = 67,["52"] = 75,["53"] = 76,["54"] = 76,["55"] = 76,["56"] = 77,["57"] = 78,["60"] = 82,["63"] = 86,["65"] = 86,["67"] = 88,["68"] = 88,["69"] = 89,["70"] = 91,["73"] = 95,["74"] = 96,["75"] = 76,["76"] = 76,["77"] = 76,["78"] = 75,["79"] = 101,["80"] = 102,["81"] = 103,["82"] = 104,["83"] = 105,["84"] = 106,["86"] = 108,["87"] = 109,["90"] = 112,["91"] = 113,["92"] = 101,["93"] = 116,["94"] = 117,["95"] = 118,["96"] = 119,["97"] = 120,["98"] = 121,["101"] = 124,["102"] = 116,["103"] = 127,["104"] = 128,["105"] = 129,["106"] = 130,["109"] = 133,["110"] = 134,["111"] = 136,["112"] = 136,["113"] = 136,["114"] = 137,["115"] = 137,["116"] = 137,["117"] = 137,["118"] = 137,["119"] = 137,["120"] = 138,["121"] = 138,["122"] = 138,["123"] = 138,["124"] = 138,["125"] = 138,["126"] = 139,["127"] = 136,["128"] = 136,["129"] = 142,["131"] = 143,["132"] = 143,["134"] = 144,["135"] = 146,["136"] = 147,["137"] = 149,["140"] = 152,["141"] = 153,["144"] = 157,["145"] = 158,["146"] = 159,["148"] = 161,["149"] = 161,["150"] = 161,["151"] = 161,["152"] = 161,["153"] = 161,["154"] = 161,["155"] = 161,["156"] = 162,["157"] = 163,["158"] = 164,["159"] = 165,["160"] = 166,["161"] = 167,["162"] = 168,["163"] = 170,["164"] = 171,["165"] = 173,["169"] = 177,["172"] = 143,["175"] = 127,["176"] = 182,["177"] = 183,["178"] = 184,["179"] = 182,["180"] = 26});
local ____exports = {}
local ____Projectile = require("solar.solar-common.tool.Projectile")
local Projectile = ____Projectile.default
local ____STimer = require("solar.solar-common.tool.STimer")
local STimer = ____STimer.default
local ____DamageType = require("solar.solar-common.constant.DamageType")
local DamageType = ____DamageType.default
local ____WeaponType = require("solar.solar-common.constant.WeaponType")
local WeaponType = ____WeaponType.default
local ____AttributeUtil = require("solar.solar-common.util.system.AttributeUtil")
local AttributeUtil = ____AttributeUtil.default
local ____SelectUtil = require("solar.solar-common.util.unit.SelectUtil")
local SelectUtil = ____SelectUtil.default
local ____MathUtil = require("solar.solar-common.util.math.MathUtil")
local MathUtil = ____MathUtil.default
local ____UnitStateUtil = require("solar.solar-common.util.unit.UnitStateUtil")
local UnitStateUtil = ____UnitStateUtil.default
____exports.default = __TS__Class()
local AutoMissileAttackController = ____exports.default
AutoMissileAttackController.name = "AutoMissileAttackController"
function AutoMissileAttackController.prototype.____constructor(self, unit, projectileModelPath, damageStateFormula, onHitTarget)
    if projectileModelPath == nil then
        projectileModelPath = "Abilities\\Weapons\\RocketMissile\\RocketMissile.mdx"
    end
    self.range = 1000
    self.extRange = 0
    self.coolDown = 1
    self.speed = 1000
    self.targetCount = 1
    self.projectileModelScale = 1
    self.colorR = 255
    self.colorG = 255
    self.colorB = 255
    self.alpha = 255
    self.damageTypeId = DamageType["s_法术"]
    self.weaponTypeID = WeaponType.T0_WHOKNOWS
    self.attackType = ATTACK_TYPE_CHAOS
    self.sTimer = __TS__New(STimer)
    self._isDestroy = false
    self._lastLaunchTime = 0
    self.unit = unit
    self.projectileModelPath = projectileModelPath
    self.damageStateFormula = damageStateFormula
    self.onHitTarget = onHitTarget
end
function AutoMissileAttackController.prototype.start(self)
    self.sTimer:start(
        0.05,
        function()
            if self._isDestroy then
                self.sTimer:destroy()
                return
            end
            if not UnitStateUtil:isAlive(self.unit) then
                return
            end
            local ____opt_0 = self.onUpdate
            if ____opt_0 ~= nil then
                ____opt_0(self, self)
            end
            local ____opt_2 = AttributeUtil:getUnitAttribute(self.unit, false)
            local autoMissileAttackSpeed = ____opt_2 and ____opt_2.autoMissileAttackSpeed or 0
            local cd = self.coolDown / (autoMissileAttackSpeed + 1)
            if _g_time - self._lastLaunchTime < cd * 1000 then
                return
            end
            self:doLaunchMissile()
            self._lastLaunchTime = _g_time
        end,
        true
    )
end
function AutoMissileAttackController.prototype.getRealRange(self)
    local range = self.range + self.extRange
    local unitAttribute = AttributeUtil:getUnitAttribute(self.unit, false)
    if unitAttribute then
        if unitAttribute.autoMissileAttackRange then
            range = range + unitAttribute.autoMissileAttackRange
        end
        if unitAttribute.autoMissileAttackRange_p then
            range = range * (1 + unitAttribute.autoMissileAttackRange_p)
        end
    end
    range = math.min(range, self.rangeMaxLimit or ____exports.default.defaultRangeMaxLimit)
    return range
end
function AutoMissileAttackController.prototype.getRealTargetCount(self)
    local targetCount = self.targetCount
    local unitAttribute = AttributeUtil:getUnitAttribute(self.unit, false)
    if unitAttribute then
        if unitAttribute.autoMissileAttackTargetCount then
            targetCount = targetCount + unitAttribute.autoMissileAttackTargetCount
        end
    end
    return targetCount
end
function AutoMissileAttackController.prototype.doLaunchMissile(self)
    local range = self:getRealRange()
    local enemyUnits = SelectUtil.getEnemyUnitsInRange(self.unit, range)
    if enemyUnits == nil or #enemyUnits == 0 then
        return
    end
    local x = GetUnitX(self.unit)
    local y = GetUnitY(self.unit)
    __TS__ArraySort(
        enemyUnits,
        function(____, a, b)
            local aD = MathUtil.distanceBetweenPoints(
                x,
                y,
                GetUnitX(a),
                GetUnitY(a)
            )
            local bD = MathUtil.distanceBetweenPoints(
                x,
                y,
                GetUnitX(b),
                GetUnitY(b)
            )
            return aD - bD
        end
    )
    local realTargetCount = self:getRealTargetCount()
    do
        local i = 0
        while i < realTargetCount and i < #enemyUnits do
            do
                local targetUnit = enemyUnits[i + 1]
                if self.onBeforeLaunchMissile then
                    local b = self:onBeforeLaunchMissile(self, targetUnit)
                    if self._isDestroy then
                        return
                    end
                    if b == false then
                        goto __continue18
                    end
                end
                local damage = 0
                if self.damageStateFormula then
                    damage = UnitStateUtil:calculateStateFormula(self.damageStateFormula, self.unit, targetUnit)
                end
                local projectile = __TS__New(
                    Projectile,
                    self.unit,
                    targetUnit,
                    damage,
                    self.projectileModelPath,
                    self.onHitTarget
                )
                projectile.speed = self.speed
                projectile.damageName = self.damageName
                projectile.projectileModelScale = self.projectileModelScale
                projectile.colorR = self.colorR
                projectile.colorG = self.colorG
                projectile.colorB = self.colorB
                projectile.alpha = self.alpha
                if self.onLaunchMissile then
                    self:onLaunchMissile(self, projectile)
                    if self._isDestroy then
                        return
                    end
                end
                projectile:start()
            end
            ::__continue18::
            i = i + 1
        end
    end
end
function AutoMissileAttackController.prototype.destroy(self)
    self._isDestroy = true
    self.sTimer:destroy()
end
AutoMissileAttackController.defaultRangeMaxLimit = 3000
return ____exports

File: src/solar/solar-common/controller/AutoMissileAttackController.ts
import Projectile from "@/Projectile";
import STimer from "@/STimer";
import DamageType from "@/DamageType";
import WeaponType from "@/WeaponType";
import AttributeUtil from "@/AttributeUtil";
import SelectUtil from "@/SelectUtil";
import MathUtil from "@/MathUtil";
import UnitStateUtil from "@/UnitStateUtil";

declare global {
    interface AppAttribute {
        /** 最大目标数量增加  */
        autoMissileAttackTargetCount?: number
        /** 射速增幅 1=1倍  */
        autoMissileAttackSpeed?: number
        /** 范围增强   */
        autoMissileAttackRange?: number
        /** 范围增强比例  0.5=增加50%射程   */
        autoMissileAttackRange_p?: number
    }
}
/**
 * 自动弹道攻击 (比如凤凰火焰)
 */
export default class AutoMissileAttackController implements Destroyable {
    static defaultRangeMaxLimit: number = 3000
    unit: unit;
    rangeMaxLimit?: number;
    range: number = 1000
    /** 扩展的范围 实际范围=range+extRange 方便组装范围控制逻辑 如A杖类型的物品 提高部分技能的范围 */
    extRange: number = 0
    coolDown: number = 1;
    /** 投射物 速度 */
    speed: number = 1000;
    targetCount: number = 1
    //伤害公式不为空的时候会自动造成公式算法的伤害
    damageStateFormula: StateFormula
    damageName?: string
    projectileModelPath: string
    /** 投射物特效模型大小缩放 */
    projectileModelScale: number = 1;
    /** 颜色rgb max=255 */
    colorR: number = 255
    colorG: number = 255
    colorB: number = 255
    /** 透明度 max=255 */
    alpha: number = 255
    damageTypeId: number = DamageType.s_法术;
    weaponTypeID: number = WeaponType.T0_WHOKNOWS;
    attackType: attacktype = ATTACK_TYPE_CHAOS
    /** 当更新时 */
    onUpdate: (controller: AutoMissileAttackController) => void
    /** 当准备发射追踪投射物时 返回false取消发射 返回true继续*/
    onBeforeLaunchMissile: (controller: AutoMissileAttackController, target: unit) => boolean
    /** 当发射追踪投射物时 可在这里修改具体的 */
    onLaunchMissile: (controller: AutoMissileAttackController, projectile: Projectile) => void
    onHitTarget: (projectile: Projectile) => void
    //
    sTimer: STimer = new STimer();
    private _isDestroy: boolean = false;
    //上次发射时间
    private _lastLaunchTime: number = 0;

    /** 任意附加数据 */
    [key: string]: any

    constructor(unit: unit, projectileModelPath: string = "Abilities\\Weapons\\RocketMissile\\RocketMissile.mdx", damageStateFormula?: StateFormula, onHitTarget?: (projectile: Projectile) => void) {
        this.unit = unit;
        this.projectileModelPath = projectileModelPath;
        this.damageStateFormula = damageStateFormula;
        this.onHitTarget = onHitTarget;
    }


    start() {
        this.sTimer.start(0.05, () => {
            if (this._isDestroy) {
                this.sTimer.destroy();
                return
            }
            //
            if (!UnitStateUtil.isAlive(this.unit)) {
                return;
            }
            //
            this.onUpdate?.(this);
            //
            let autoMissileAttackSpeed = AttributeUtil.getUnitAttribute(this.unit, false)?.autoMissileAttackSpeed || 0
            let cd = this.coolDown / (autoMissileAttackSpeed + 1)
            //判断是否在cd中
            if (_g_time - this._lastLaunchTime < (cd * 1000)) {
                return;
            }
            //准备发射
            this.doLaunchMissile();
            this._lastLaunchTime = _g_time;
        }, true)

    }

    getRealRange(): number {
        let range = this.range + this.extRange;
        let unitAttribute = AttributeUtil.getUnitAttribute(this.unit, false);
        if (unitAttribute) {
            if (unitAttribute.autoMissileAttackRange) {
                range += unitAttribute.autoMissileAttackRange;
            }
            if (unitAttribute.autoMissileAttackRange_p) {
                range *= (1 + unitAttribute.autoMissileAttackRange_p);
            }
        }
        range = math.min(range, this.rangeMaxLimit || AutoMissileAttackController.defaultRangeMaxLimit)
        return range;
    }

    getRealTargetCount(): number {
        let targetCount = this.targetCount;
        let unitAttribute = AttributeUtil.getUnitAttribute(this.unit, false);
        if (unitAttribute) {
            if (unitAttribute.autoMissileAttackTargetCount) {
                targetCount += unitAttribute.autoMissileAttackTargetCount;
            }
        }
        return targetCount;
    }

    private doLaunchMissile() {
        let range = this.getRealRange();
        let enemyUnits = SelectUtil.getEnemyUnitsInRange(this.unit, range);
        if (enemyUnits == null || enemyUnits.length == 0) {
            return
        }
        let x = GetUnitX(this.unit)
        let y = GetUnitY(this.unit)
        //按距离 由近到远
        enemyUnits.sort((a, b) => {
            let aD: number = MathUtil.distanceBetweenPoints(x, y, GetUnitX(a), GetUnitY(a));
            let bD: number = MathUtil.distanceBetweenPoints(x, y, GetUnitX(b), GetUnitY(b));
            return aD - bD;
        });

        let realTargetCount = this.getRealTargetCount();
        for (let i = 0; i < realTargetCount && i < enemyUnits.length; i++) {
            let targetUnit = enemyUnits[i];
            //准备发射导弹
            if (this.onBeforeLaunchMissile) {
                let b = this.onBeforeLaunchMissile(this, targetUnit);
                //检查是否在监听里销毁了此类
                if (this._isDestroy) {
                    return;
                }
                if (b == false) {
                    continue;
                }
            }
            //发射导弹
            let damage = 0
            if (this.damageStateFormula) {
                damage = UnitStateUtil.calculateStateFormula(this.damageStateFormula, this.unit, targetUnit)
            }
            let projectile: Projectile = new Projectile(this.unit, targetUnit, damage, this.projectileModelPath, this.onHitTarget);
            projectile.speed = this.speed
            projectile.damageName = this.damageName
            projectile.projectileModelScale = this.projectileModelScale
            projectile.colorR = this.colorR
            projectile.colorG = this.colorG
            projectile.colorB = this.colorB
            projectile.alpha = this.alpha
            //
            if (this.onLaunchMissile) {
                this.onLaunchMissile(this, projectile);
                //检查是否在监听里销毁了此类
                if (this._isDestroy) {
                    return;
                }
            }
            projectile.start()
        }
    }


    destroy() {
        this._isDestroy = true;
        this.sTimer.destroy();
    }


}